File: packages/core/src/changeset.ts
import { Schema } from "effect"

export class FileChange extends Schema.Class<FileChange>("FileChange")({
  path: Schema.String,
  kind: Schema.Literals(["add", "modify", "delete", "move"]),
  fromPath: Schema.optionalKey(Schema.String),
  oldSnapshot: Schema.optionalKey(Schema.String),
  newSnapshot: Schema.optionalKey(Schema.String),
  additions: Schema.Number,
  deletions: Schema.Number,
  wholeFile: Schema.Boolean,
  toolCallId: Schema.String,
}) {}

export class ChangeSet extends Schema.Class<ChangeSet>("ChangeSet")({
  sessionId: Schema.String,
  turnId: Schema.String,
  title: Schema.String,
  files: Schema.Array(FileChange),
  createdAt: Schema.Number,
}) {}

export const SNAPSHOT_FILE_CAP = 32
export const SNAPSHOT_BYTE_CAP = 16 * 1024 * 1024
export const MISSING_SNAPSHOT_REASON = "missing snapshot"
export const REGION_ONLY_REASON = "region only"

export class FileChangeRecord extends Schema.Class<FileChangeRecord>("FileChangeRecord")({
  path: Schema.String,
  kind: Schema.Literals(["add", "modify", "delete", "move"]),
  fromPath: Schema.optionalKey(Schema.String),
  additions: Schema.Number,
  deletions: Schema.Number,
  wholeFile: Schema.Boolean,
  toolCallId: Schema.String,
  snapshotStored: Schema.Boolean,
  snapshotBytes: Schema.Number,
  undoDisabledReason: Schema.optionalKey(Schema.String),
}) {}

export class ChangeSetRecord extends Schema.Class<ChangeSetRecord>("ChangeSetRecord")({
  sessionId: Schema.String,
  turnId: Schema.String,
  title: Schema.String,
  files: Schema.Array(FileChangeRecord),
  createdAt: Schema.Number,
}) {}

export const ChangeSetIndex = Schema.Array(ChangeSetRecord)
export type ChangeSetIndex = typeof ChangeSetIndex.Type

export const canStoreSnapshot = (
  storedFiles: number,
  storedBytes: number,
  extraBytes: number,
): boolean => storedFiles < SNAPSHOT_FILE_CAP && storedBytes + extraBytes <= SNAPSHOT_BYTE_CAP

export const storedSnapshotBudget = (
  sets: ReadonlyArray<ChangeSetRecord>,
): { readonly files: number; readonly bytes: number } => {
  let files = 0
  let bytes = 0
  for (const set of sets) {
    for (const file of set.files) {
      if (!file.snapshotStored) continue
      files += 1
      bytes += file.snapshotBytes
    }
  }
  return { files, bytes }
}

export const changeSetLineStats = (
  files: ReadonlyArray<{ readonly additions: number; readonly deletions: number }>,
): { readonly additions: number; readonly deletions: number } =>
  files.reduce(
    (acc, file) => ({
      additions: acc.additions + file.additions,
      deletions: acc.deletions + file.deletions,
    }),
    { additions: 0, deletions: 0 },
  )

export const formatLineStats = (additions: number, deletions: number): string =>
  `+${additions}/\u2212${deletions}`

export const turnTitleFromPrompt = (text: string): string => {
  const line = text.split(/\r?\n/).map((row) => row.trim()).find((row) => row.length > 0)
  if (line === undefined) return "Untitled"
  return line.length > 80 ? `${line.slice(0, 77)}...` : line
}

export const fileChangeFromRecord = (
  record: FileChangeRecord,
  snapshots?: { readonly oldText?: string; readonly newText?: string },
): FileChange =>
  new FileChange({
    path: record.path,
    kind: record.kind,
    additions: record.additions,
    deletions: record.deletions,
    wholeFile: record.wholeFile,
    toolCallId: record.toolCallId,
    ...(record.fromPath !== undefined ? { fromPath: record.fromPath } : {}),
    ...(snapshots?.oldText !== undefined ? { oldSnapshot: snapshots.oldText } : {}),
    ...(snapshots?.newText !== undefined ? { newSnapshot: snapshots.newText } : {}),
  })

export const undoDisabledReasonFor = (
  change: { readonly kind: FileChange["kind"]; readonly wholeFile: boolean },
  snapshotStored: boolean,
): string | undefined => {
  if (!snapshotStored) return MISSING_SNAPSHOT_REASON
  if (change.kind === "modify" && !change.wholeFile) return REGION_ONLY_REASON
  return undefined
}

export const shouldKeepExistingFileChange = (
  existing: { readonly wholeFile: boolean; readonly snapshotStored: boolean },
  incoming: { readonly wholeFile: boolean; readonly regionStandIn?: boolean },
): boolean =>
  existing.snapshotStored
  && existing.wholeFile
  && (!incoming.wholeFile || incoming.regionStandIn === true)

export const recordFromFileChange = (
  change: FileChange,
  stored: { readonly snapshotStored: boolean; readonly snapshotBytes: number },
): FileChangeRecord => {
  const undoDisabledReason = undoDisabledReasonFor(change, stored.snapshotStored)
  return new FileChangeRecord({
    path: change.path,
    kind: change.kind,
    additions: change.additions,
    deletions: change.deletions,
    wholeFile: change.wholeFile,
    toolCallId: change.toolCallId,
    snapshotStored: stored.snapshotStored,
    snapshotBytes: stored.snapshotBytes,
    ...(change.fromPath !== undefined ? { fromPath: change.fromPath } : {}),
    ...(undoDisabledReason !== undefined ? { undoDisabledReason } : {}),
  })
}

export type UndoPlan =
  | { readonly _tag: "replace"; readonly path: string; readonly text: string }
  | { readonly _tag: "create"; readonly path: string; readonly text: string }
  | { readonly _tag: "delete"; readonly path: string; readonly confirmIfDirty: boolean }
  | {
    readonly _tag: "moveReverse"
    readonly fromPath: string
    readonly toPath: string
    readonly oldText: string
  }
  | { readonly _tag: "disabled"; readonly reason: string }

export const keepFile = (set: ChangeSet, path: string): ChangeSet =>
  new ChangeSet({
    ...set,
    files: set.files.filter((file) => file.path !== path),
  })

export const keepAll = (set: ChangeSet): ChangeSet => new ChangeSet({ ...set, files: [] })

const splitLines = (text: string): ReadonlyArray<string> => text.split(/\r?\n/)

export const lineDiffStats = (
  oldText: string,
  newText: string,
): { readonly additions: number; readonly deletions: number } => {
  const oldLines = splitLines(oldText)
  const newLines = splitLines(newText)
  const lcs = longestCommonSubsequence(oldLines, newLines)
  return {
    deletions: oldLines.length - lcs,
    additions: newLines.length - lcs,
  }
}

const longestCommonSubsequence = (
  a: ReadonlyArray<string>,
  b: ReadonlyArray<string>,
): number => {
  const n = a.length
  const m = b.length
  if (n === 0 || m === 0) return 0
  let prev = new Array<number>(m + 1).fill(0)
  let curr = new Array<number>(m + 1).fill(0)
  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      curr[j] = a[i - 1] === b[j - 1]
        ? (prev[j - 1] ?? 0) + 1
        : Math.max(prev[j] ?? 0, curr[j - 1] ?? 0)
    }
    const tmp = prev
    prev = curr
    curr = tmp
    curr.fill(0)
  }
  return prev[m] ?? 0
}

export const undoPlan = (change: FileChange, diskNow?: string): UndoPlan => {
  switch (change.kind) {
    case "modify": {
      if (!change.wholeFile) return { _tag: "disabled", reason: REGION_ONLY_REASON }
      if (change.oldSnapshot === undefined) return { _tag: "disabled", reason: "missing snapshot" }
      return { _tag: "replace", path: change.path, text: change.oldSnapshot }
    }
    case "delete": {
      if (change.oldSnapshot === undefined) return { _tag: "disabled", reason: "missing snapshot" }
      if (diskNow !== undefined && diskNow !== change.oldSnapshot) {
        return { _tag: "disabled", reason: "path exists with different content" }
      }
      return { _tag: "create", path: change.path, text: change.oldSnapshot }
    }
    case "add": {
      const dirty = diskNow !== undefined
        && change.newSnapshot !== undefined
        && diskNow !== change.newSnapshot
      return { _tag: "delete", path: change.path, confirmIfDirty: dirty }
    }
    case "move": {
      if (change.fromPath === undefined) return { _tag: "disabled", reason: "move target unknown" }
      if (change.oldSnapshot === undefined) return { _tag: "disabled", reason: "missing snapshot" }
      return {
        _tag: "moveReverse",
        fromPath: change.fromPath,
        toPath: change.path,
        oldText: change.oldSnapshot,
      }
    }
  }
}

export const applyUndoToSnapshots = (change: FileChange): string | undefined => {
  const plan = undoPlan(change)
  switch (plan._tag) {
    case "replace":
    case "create":
    case "moveReverse":
      return change.oldSnapshot
    case "delete":
      return undefined
    case "disabled":
      return change.oldSnapshot
  }
}

export type UndoMutation =
  | { readonly _tag: "replace"; readonly path: string; readonly text: string }
  | { readonly _tag: "create"; readonly path: string; readonly text: string }
  | { readonly _tag: "delete"; readonly path: string }

export type UndoResolution =
  | { readonly _tag: "apply"; readonly mutations: ReadonlyArray<UndoMutation> }
  | { readonly _tag: "disabled"; readonly reason: string }
  | { readonly _tag: "cancelled" }

export const mutationsFromUndoPlan = (
  plan: Exclude<UndoPlan, { readonly _tag: "disabled" }>,
): ReadonlyArray<UndoMutation> => {
  switch (plan._tag) {
    case "replace":
      return [{ _tag: "replace", path: plan.path, text: plan.text }]
    case "create":
      return [{ _tag: "create", path: plan.path, text: plan.text }]
    case "delete":
      return [{ _tag: "delete", path: plan.path }]
    case "moveReverse":
      return [
        { _tag: "create", path: plan.fromPath, text: plan.oldText },
        { _tag: "delete", path: plan.toPath },
      ]
  }
}

export const resolveUndo = async (
  change: FileChange,
  diskNow: string | undefined,
  confirmDirty: () => Promise<boolean>,
): Promise<UndoResolution> => {
  const plan = undoPlan(change, diskNow)
  if (plan._tag === "disabled") return plan
  if (plan._tag === "delete" && plan.confirmIfDirty) {
    if (!(await confirmDirty())) return { _tag: "cancelled" }
  }
  return { _tag: "apply", mutations: mutationsFromUndoPlan(plan) }
}

File: beard/core/src/main/scala/groksbeard/core/ChangeSet.scala
package groksbeard.core

enum ChangeKind:
  case Add, Modify, Delete, Move

object ChangeKind:
  def wire(kind: ChangeKind): String =
    kind match
      case ChangeKind.Add    => "add"
      case ChangeKind.Modify => "modify"
      case ChangeKind.Delete => "delete"
      case ChangeKind.Move   => "move"

  def fromWire(value: String): ChangeKind =
    value match
      case "add"    => ChangeKind.Add
      case "delete" => ChangeKind.Delete
      case "move"   => ChangeKind.Move
      case _        => ChangeKind.Modify
end ChangeKind

final case class FileChange(
    path: String,
    kind: ChangeKind,
    additions: Int,
    deletions: Int,
    wholeFile: Boolean,
    toolCallId: String,
    fromPath: Option[String] = None,
    oldSnapshot: Option[String] = None,
    newSnapshot: Option[String] = None,
    undoDisabled: Option[String] = None,
)

final case class ChangeSet(
    sessionId: String,
    turnId: String,
    title: String,
    files: List[FileChange],
    createdAt: Long,
)

enum UndoPlan:
  case Replace(path: String, text: String)
  case Create(path: String, text: String)
  case Delete(path: String, confirmIfDirty: Boolean)
  case MoveReverse(fromPath: String, toPath: String, oldText: String)
  case Disabled(reason: String)

enum UndoMutation:
  case Replace(path: String, text: String)
  case Create(path: String, text: String)
  case Delete(path: String)

enum UndoResolution:
  case Apply(mutations: List[UndoMutation])
  case Disabled(reason: String)
  case Cancelled

object ChangeSet:
  val SnapshotFileCap: Int        = 32
  val SnapshotByteCap: Int        = 16 * 1024 * 1024
  val MissingSnapshot: String     = "missing snapshot"
  val RegionOnly: String          = "region only"
  val MoveTargetUnknown: String   = "move target unknown"
  val PathExistsDifferent: String = "path exists with different content"

  def keepFile(set: ChangeSet, path: String): ChangeSet =
    set.copy(files = set.files.filterNot(_.path == path))

  def keepAll(set: ChangeSet): ChangeSet =
    set.copy(files = Nil)

  def lineStats(files: List[FileChange]): (Int, Int) =
    files.foldLeft((0, 0)) { case ((add, del), f) => (add + f.additions, del + f.deletions) }

  def formatStats(additions: Int, deletions: Int): String =
    s"+$additions/-$deletions"

  def turnTitle(text: String): String =
    text.split("\r?\n").iterator.map(_.trim).find(_.nonEmpty) match
      case None                            => "Untitled"
      case Some(line) if line.length <= 80 => line
      case Some(line)                      => s"${line.take(77)}..."

  def canStoreSnapshot(storedFiles: Int, storedBytes: Int, extraBytes: Int): Boolean =
    storedFiles < SnapshotFileCap && storedBytes + extraBytes <= SnapshotByteCap

  def snapshotBytes(change: FileChange): Int =
    change.oldSnapshot.map(Utf8.byteLength).getOrElse(0) +
      change.newSnapshot.map(Utf8.byteLength).getOrElse(0)

  def undoDisabledFor(kind: ChangeKind, wholeFile: Boolean, snapshotStored: Boolean): Option[String] =
    if !snapshotStored then Some(MissingSnapshot)
    else if kind == ChangeKind.Modify && !wholeFile then Some(RegionOnly)
    else None

  def shouldKeepExisting(
      existingWholeFile: Boolean,
      existingStored: Boolean,
      incomingWholeFile: Boolean,
      regionStandIn: Boolean,
  ): Boolean =
    existingStored && existingWholeFile && (!incomingWholeFile || regionStandIn)

  def lineDiffStats(oldText: String, newText: String): (Int, Int) =
    val oldLines = splitLines(oldText)
    val newLines = splitLines(newText)
    val lcs      = longestCommonSubsequence(oldLines, newLines)
    (newLines.length - lcs, oldLines.length - lcs)

  def undoPlan(change: FileChange, diskNow: Option[String] = None): UndoPlan =
    change.kind match
      case ChangeKind.Modify =>
        if !change.wholeFile then UndoPlan.Disabled(RegionOnly)
        else
          change.oldSnapshot match
            case None    => UndoPlan.Disabled(MissingSnapshot)
            case Some(s) => UndoPlan.Replace(change.path, s)
      case ChangeKind.Delete =>
        change.oldSnapshot match
          case None    => UndoPlan.Disabled(MissingSnapshot)
          case Some(s) =>
            diskNow match
              case Some(now) if now != s => UndoPlan.Disabled(PathExistsDifferent)
              case _                     => UndoPlan.Create(change.path, s)
      case ChangeKind.Add =>
        val dirty = diskNow.exists(now => change.newSnapshot.exists(_ != now))
        UndoPlan.Delete(change.path, confirmIfDirty = dirty)
      case ChangeKind.Move =>
        (change.fromPath, change.oldSnapshot) match
          case (None, _)             => UndoPlan.Disabled(MoveTargetUnknown)
          case (_, None)             => UndoPlan.Disabled(MissingSnapshot)
          case (Some(from), Some(s)) =>
            UndoPlan.MoveReverse(from, change.path, s)

  def mutationsFrom(plan: UndoPlan): List[UndoMutation] =
    plan match
      case UndoPlan.Replace(path, text)            => List(UndoMutation.Replace(path, text))
      case UndoPlan.Create(path, text)             => List(UndoMutation.Create(path, text))
      case UndoPlan.Delete(path, _)                => List(UndoMutation.Delete(path))
      case UndoPlan.MoveReverse(from, to, oldText) =>
        List(UndoMutation.Create(from, oldText), UndoMutation.Delete(to))
      case UndoPlan.Disabled(_) => Nil

  def resolveUndo(change: FileChange, diskNow: Option[String], confirmDirty: Boolean): UndoResolution =
    undoPlan(change, diskNow) match
      case UndoPlan.Disabled(reason)                                     => UndoResolution.Disabled(reason)
      case plan: UndoPlan.Delete if plan.confirmIfDirty && !confirmDirty =>
        UndoResolution.Cancelled
      case plan => UndoResolution.Apply(mutationsFrom(plan))

  def applyUndoToSnapshots(change: FileChange): Option[String] =
    undoPlan(change) match
      case UndoPlan.Delete(_, _) => None
      case _                     => change.oldSnapshot

  def toView(change: FileChange): ChangeFileView =
    ChangeFileView(
      path = change.path,
      kind = ChangeKind.wire(change.kind),
      additions = change.additions,
      deletions = change.deletions,
      wholeFile = change.wholeFile,
      undoDisabled = change.undoDisabled,
    )

  def summaryOf(files: List[FileChange]): ChangesSummary =
    val (add, del) = lineStats(files)
    ChangesSummary(files.size, add, del, files.map(toView))

  private def splitLines(text: String): Array[String] =
    text.split("\r?\n", -1)

  private def longestCommonSubsequence(a: Array[String], b: Array[String]): Int =
    val n = a.length
    val m = b.length
    if n == 0 || m == 0 then 0
    else
      var prev = Array.fill(m + 1)(0)
      var curr = Array.fill(m + 1)(0)
      var i    = 1
      while i <= n do
        var j = 1
        while j <= m do
          curr(j) =
            if a(i - 1) == b(j - 1) then prev(j - 1) + 1
            else math.max(prev(j), curr(j - 1))
          j += 1
        val tmp = prev
        prev = curr
        curr = tmp
        java.util.Arrays.fill(curr, 0)
        i += 1
      end while
      prev(m)
    end if
  end longestCommonSubsequence
end ChangeSet

final class ChangeStore:
  private var sets: List[ChangeSet] = Nil

  def list: List[ChangeSet] = sets

  def pending: List[FileChange] = sets.flatMap(_.files)

  def summary: ChangesSummary = ChangeSet.summaryOf(pending)

  def get(path: String): Option[FileChange] =
    pending.find(_.path == path)

  def ingest(sessionId: String, turnId: String, title: String, incoming: List[FileChange]): Unit =
    if incoming.isEmpty then ()
    else
      val idx = sets.indexWhere(s => s.sessionId == sessionId && s.turnId == turnId)
      val now = System.currentTimeMillis()
      if idx < 0 then sets = sets :+ ChangeSet(sessionId, turnId, title, incoming.map(withBudget), now)
      else
        val current = sets(idx)
        val merged  = incoming.foldLeft(current.files)(mergeFile)
        sets = sets.updated(idx, current.copy(title = title, files = merged))

  def keep(path: String): Unit =
    sets = sets.map(ChangeSet.keepFile(_, path)).filter(_.files.nonEmpty)

  def keepAll(): Unit =
    sets = Nil

  def drop(path: String): Unit = keep(path)

  private def mergeFile(files: List[FileChange], incoming: FileChange): List[FileChange] =
    val idx = files.indexWhere(_.path == incoming.path)
    if idx < 0 then files :+ withBudget(incoming)
    else
      val existing = files(idx)
      if ChangeSet.shouldKeepExisting(
          existing.wholeFile,
          existing.oldSnapshot.isDefined || existing.newSnapshot.isDefined,
          incoming.wholeFile,
          regionStandIn = incoming.wholeFile && incoming.undoDisabled.contains(ChangeSet.RegionOnly),
        )
      then files
      else files.updated(idx, withBudget(incoming))
    end if
  end mergeFile

  private def withBudget(change: FileChange): FileChange =
    val stored = sets.flatMap(_.files)
    val files  = stored.count(f => f.oldSnapshot.isDefined || f.newSnapshot.isDefined)
    val bytes  = stored.map(ChangeSet.snapshotBytes).sum
    val extra  = ChangeSet.snapshotBytes(change)
    if ChangeSet.canStoreSnapshot(files, bytes, extra) then
      change.copy(undoDisabled = ChangeSet.undoDisabledFor(change.kind, change.wholeFile, snapshotStored = true))
    else
      change.copy(
        oldSnapshot = None,
        newSnapshot = None,
        undoDisabled = Some(ChangeSet.MissingSnapshot),
      )
  end withBudget
end ChangeStore
